modules/layers/src/solid-polygon-layer/polygon-tesselator.js

```diff
@@ -132,40 +133,48 @@ export default class PolygonTesselator extends Tesselator {
   }
 
   // Flatten out all the vertices of all the sub subPolygons
-  _updatePositions(polygon, {vertexStart}) {
+  _updatePositions(polygon, {vertexStart, geometrySize}) {
     const {
       attributes: {positions, positions64xyLow, vertexValid},
-      fp64
+      fp64,
+      positionSize
     } = this;
 
     let i = vertexStart;
-    polygon.forEach(loop => {
-      loop.forEach(vertex => {
-        // eslint-disable-line
-        const x = vertex[0];
-        const y = vertex[1];
-        const z = vertex[2] || 0;
-
-        positions[i * 3] = x;
-        positions[i * 3 + 1] = y;
-        positions[i * 3 + 2] = z;
-        if (fp64) {
-          positions64xyLow[i * 2] = fp64LowPart(x);
-          positions64xyLow[i * 2 + 1] = fp64LowPart(y);
-        }
-        vertexValid[i] = 1;
-        i++;
-      });
-      /* We are reusing the some buffer for `nextPositions` by offsetting one vertex
-       * to the left. As a result,
-       * the last vertex of each loop overlaps with the first vertex of the next loop.
-       * `vertexValid` is used to mark the end of each loop so we don't draw these
-       * segments:
-        positions      A0 A1 A2 A3 A4 B0 B1 B2 C0 ...
-        nextPositions  A1 A2 A3 A4 B0 B1 B2 C0 C1 ...
-        vertexValid    1  1  1  1  0  1  1  0  1 ...
-       */
-      vertexValid[i - 1] = 0;
-    });
+    const {positions: polygonPositions, loopStartIndices} = polygon;
+
+    for (let j = 0; j < geometrySize; j++) {
+      const x = polygonPositions[j * positionSize];
+      const y = polygonPositions[j * positionSize + 1];
+      const z = positionSize > 2 ? polygonPositions[j * positionSize + 2] : 0;
+
+      positions[i * 3] = x;
+      positions[i * 3 + 1] = y;
+      positions[i * 3 + 2] = z;
+      if (fp64) {
+        positions64xyLow[i * 2] = fp64LowPart(x);
+        positions64xyLow[i * 2 + 1] = fp64LowPart(y);
+      }
+      vertexValid[i] = 1;
+      i++;
+    }
+
+    /* We are reusing the some buffer for `nextPositions` by offsetting one vertex
+     * to the left. As a result,
+     * the last vertex of each loop overlaps with the first vertex of the next loop.
+     * `vertexValid` is used to mark the end of each loop so we don't draw these
+     * segments:
+      positions      A0 A1 A2 A3 A4 B0 B1 B2 C0 ...
+      nextPositions  A1 A2 A3 A4 B0 B1 B2 C0 C1 ...
+      vertexValid    1  1  1  1  0  1  1  0  1 ...
+     */
+    if (loopStartIndices) {
+      for (let j = 1; j <= loopStartIndices.length; j++) {
+        const loopStartIndex = loopStartIndices[j] || geometrySize;
```

Comment:
Can we add a line of comment for line 173?
Is it to prevent loopStartIndices = [0, 0, ...]?
In which case this two zeros would happen? Like for a specific polygon path format?